### PR TITLE
metal : reusing llama.cpp logging

### PR DESCRIPTION
**Pull request by Ricardicus**
I wanted to silence some of the outputs and found out that some
came from llama.cpp and some from metal.m on my mac. I saw the "TODO" and
thought I might chime in here with this. Perhaps we should regard the log level also,
and be able to set verbosity via command line arguments? Let me know what you think.

### CODE REVIEW

**Comment by Cebtenzzre** [comment]
The cmake build is failing on macOS:
```
[ 91%] Linking CXX executable ../../bin/metal
Undefined symbols for architecture x86_64:
  "_llama_log", referenced from:
      _ggml_metal_init in ggml-metal.m.o
      _ggml_metal_free in ggml-metal.m.o
      _ggml_metal_host_malloc in ggml-metal.m.o
      _ggml_metal_add_buffer in ggml-metal.m.o
      _ggml_metal_set_tensor in ggml-metal.m.o
      _ggml_metal_get_tensor in ggml-metal.m.o
      _ggml_metal_graph_find_concurrency in ggml-metal.m.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bin/metal] Error 1
make[1]: *** [examples/metal/CMakeFiles/metal.dir/all] Error 2
```

**Comment by Ricardicus** [comment]
Ok, sorry I missed that. Fixed it

**Comment by ggerganov** [comment]
The change in this PR is not OK because it couples `ggml` with `llama.cpp`

I didn't write a detailed explanation in the TODO, but what I meant is to implement a way to pass a log callback and have `llama.cpp` or other projects be able to provide their callback to the metal backend.

**Comment by Ricardicus** [comment]
Ok! I see. I can try to do that, and decouple the two again.

**Comment by Ricardicus** [comment]
I decoupled it. Introducing a log function setter in ggml-metal that is called from llama.cpp that points to the internal one.

**Comment by Ricardicus** [comment]
Fixed the trailing whitespace pipeline error.

**Comment by alitariq4589** [comment]
Can one of the admins verify this patch?

**Comment by Ricardicus** [comment]
I resolved a conflict that had appeared.

**Comment by Ricardicus** [comment]
I decoupled it more now. Since log level is not really used yet I took the liberty to move that into a definition in ggml.h instead. My reasoning was that llama.cpp is already depending on ggml.h, so in order to synchronize the two with a callback I needed a function signature that could be passed in both metal and llama.cpp translation units. Maybe it would be easier to work with ints and macros, but the idea of an enum also makes sense.

File: ggml-metal.m [hunk 1]
```diff
@@ -118,8 +121,37 @@ @interface GGMLMetalClass : NSObject
...
+    } else {
+        char* buffer2 = malloc(len+1);
+        vsnprintf(buffer2, len+1, format, args);
+        buffer2[len] = 0;
+        ggml_metal_log_callback(level, buffer2, ggml_metal_log_user_data);
+        free(buffer2);
+    }
+    va_end(args);
+  }
+}
```
Comment (ggerganov) [inline ggml-metal.m:149]: Fix indentation - 4 spaces

**Review by ggerganov — COMMENTED** [review]
> I decoupled it more now.

Yes - this is the way :) I think it OK now - maybe move the `llama_log_callback` typedef in `ggml.h` so we can reuse it:

```c
typedef void (*ggml_log_callback)(enum ggml_log_level level, const char * text, void * user_data);
```

**Comment by Ricardicus** [comment]
Ok nice! I fixed a typedef for log callbacks in ggml.h now :) just a question, I use clang-format for formatting. Is there any tools like that used in this project?

**Comment by Ricardicus** [comment]
I fixed a typedef for log callbacks in ggml.h now :)

**Review by ggerganov — APPROVED** [review]

**Pull request merged**